Add initial docs (#18)
yuvipanda/jupyterhub-ssh: README.md

> # JupyterHubSSH

Reviewer: ```suggestion
# JupyterHub SSH
```

Perhaps like this?

[![Documentation build status](https://img.shields.io/readthedocs/jupyterhub?logo=read-the-docs)](https://jupyterhub-ssh.readthedocs.io/en/latest/)

SSH Access to JupyterHubs, independent of the way there were deployed.

## Development Status
This project is under active develpoment :tada:, so expect a few changes along the way.

## Technical Overview

JupyterHubSSH is made up of two main components:

> - a Notebook SSH server that maps a SSH connection to a Notebook server on a JupyterHub.

Reviewer 2: ```suggestion
- an SSH server that maps a SSH connection to a Notebook server on a JupyterHub.
```

- a [Terminado](https://github.com/jupyter/terminado) client that knows how to connect and communicate to a Jupyter terminal.

![Overview](https://raw.githubusercontent.com/yuvipanda/jupyterhub-ssh/main/docs/source/_static/images/technical-overview.png)

> 

Reviewer: Some note about the role of the helm-chart in this repo would be nice to have as part of this section.

Author: To be honest, I missed the helm-chart part in the codebase :sweat_smile: I'll look into it and add it here once I get an understanding of it :D. Thanks @consideRatio :heart:

## Installation

1. Clone the repo and install the jupyterhub-ssh package:
``` bash
$ git clone https://github.com/yuvipanda/jupyterhub-ssh.git
$ cd jupyterhub-ssh

> $ pip install .

Reviewer: ```suggestion
$ pip install -e .
```

Do we want to suggest an editable install here, so changes to the source will be reflected after install?

Author: Why not? :+1:

```
1. Or install the package directly:

> ``` bash
> pip install git+https://github.com/yuvipanda/jupyterhub-ssh.git
> ```

Reviewer: ```suggestion
   ``` bash
   pip install git+https://github.com/yuvipanda/jupyterhub-ssh.git
   ```
```

2. Create the config file:

> ```bash
> $ touch jupyterhub_ssh_config.py
> ```

Reviewer: ```suggestion
   ```bash
   $ touch jupyterhub_ssh_config.py
   ```
```

> 3. Put in the config file at least the following two config options:
> * `c.JupyterHubSSH.hub_url`: URL of JupyterHub to connect to.
> * `c.JupyterHubSSH.host_key_path`: Path to host's private SSH Key.

Reviewer: ```suggestion
3. Put in the config file at least the following two config options:
   * `c.JupyterHubSSH.hub_url`: URL of JupyterHub to connect to.
   * `c.JupyterHubSSH.host_key_path`: Path to host's private SSH Key.
```

	More configuration options can be found in the docs [here](https://jupyterhub-ssh.readthedocs.io/en/latest/api/index.html#module-jupyterhub_ssh).

5. Start the JupyterHubSSH app from the directory where the config file

> `jupyterhub_ssh_config.py` is located:
> ```bash
> $ python -m jupyterhub_ssh
> ```

Reviewer: ```suggestion
   `jupyterhub_ssh_config.py` is located:
   ```bash
   $ python -m jupyterhub_ssh
   ```
```

## Instructions on how to use it

1. Login into your JupyterHub and go to `https://<hub-address>/hub/token`.
2. Copy the token from JupyterHub.
3. SSH into JupyterHub:

> ```bash
> $ ssh <username-you-used>@<hub-address>
> ```

Reviewer: ```suggestion
   ```bash
   $ ssh <username-you-used>@<hub-address>
   ```
```

# JupyterHub SSH and SFTP

[![Documentation build status](https://img.shields.io/readthedocs/jupyterhub?logo=read-the-docs)](https://jupyterhub-ssh.readthedocs.io/en/latest/)

> Access through [SSH](https://www.ssh.com/ssh) any JupyterHub, regardless how it was deployed and easily transfer files through [SFTP](https://www.ssh.com/ssh/sftp).

Reviewer: ```suggestion
With a JupyterHub [SSH](https://www.ssh.com/ssh) server deployed, you can start and access your JupyterHub user environment through SSH. With a JupyterHub [SFTP](https://www.ssh.com/ssh/sftp) server deployed alongside the JupyterHub's user storage, you can use SFTP to work against your JupyterHub user's home directory. Access to these services are authenticated using an access token acquired from your JupyterHub's user interface under `/hub/token`.
```

## Development Status
This project is under active develpoment :tada:, so expect a few changes along the way.

## Technical Overview

> The JupyterHubSSH service is used for SSH access to your hub. `JupyterHubSSH` is made up of two main components:

Reviewer: ```suggestion
The JupyterHub SSH service provides SSH access to your JupyterHub's user environment. JupyterHub SSH is made up of two main components:
```

	$ python -m jupyterhub_ssh
	```

### Kuberbetes based deployment
If your JupyterHub was deployed using Kubernetes, you can use the Helm charts available in this repo to deploy JupyterHub SSH & SFTP
directly into your Kubernetes cluster.

> 	```bash
> 	# Let helm the command line tool know about a Helm chart repository
> 	# that we decide to name jupyterhub.
> 	$ helm repo add jupyterhub https://yuvipanda.github.io/jupyterhub-ssh/
> 	$ helm repo update
> 
> 	# Simplified example on how to install a Helm chart from a Helm chart repository
> 	# named jupyterhub-ssh. See the Helm chart's documentation for additional details
> 	# required.
> 	$ helm install jupyterhub-ssh/<helm chart name> --version <helm chart version>
> 	```

Reviewer: ```suggestion
	```bash
	# Let helm the command line tool know about a Helm chart repository
	# that we decide to name jupyterhub.
	$ helm repo add jupyterhub-ssh https://yuvipanda.github.io/jupyterhub-ssh/
	$ helm repo update

	# Simplified example on how to install a Helm chart from a Helm chart repository
	# named jupyterhub-ssh. See the Helm chart's documentation for additional details
	# required.
	$ helm install jupyterhub-ssh/jupyterhub-ssh --version <helm chart version> --set hubUrl=https://jupyter.example.org --set-file hostKey=<path to a private SSH key>
	```
```

I think this is about right. Most people have a helm repository for charts named JupyterHub already, so reusing jupyterhub would be problematic, so let's call this Helm chart repository (in this repo's gh-pages) to jupyterhub-ssh for now.

I also filled in the helm chart name, which is well known and won't change, to jupyterhub-ssh.

I also filled in the two variables I know that is strictly required currently.

@yuvipanda btw, I think we need too have SFTP disabled by default for this Helm chart as we also require a PVC name for it.

## How to use it

### How to SSH
1. Login into your JupyterHub and go to `https://<hub-address>/hub/token`.
2. Copy the token from JupyterHub.
3. SSH into JupyterHub:
	```bash
	$ ssh <username-you-used>@<hub-address>
	```
4. Enter the token received from JupyterHub as a password.
5. TADA :tada: Now you have an interactive terminal! You can do anything you would generally interactively do via ssh: run editors,
fully interactive programs, use the commandline, etc. Some features like non-interactive command running, tunneling, etc are currently
unavailable.

### How to SFTP
1. Login into your JupyterHub and go to `https://<hub-address>/hub/token`.
2. Copy the token from JupyterHub.
3. Transfer file into Jupyterhub:
	* Using the `sftp` command:
		```bash

> 		$ sftp -oPort=2222 <hub-username>@<hub-address>

Reviewer: I think -o flag used to declare a ssh_option, and the option Port=2222, can be omitted, as the SSH entry to the SFTP server is exposed publicly on port 22, even though the container expose it at port 2222. But, I guess this depends a bit.

It will always depend on how the SFTP server is exposed, but I suggest to assume port 22 like would be correct if SFTP is deployed as part of the Helm chart.

```suggestion
		$ sftp <hub-username>@<hub-address>
```

		```

> 	* Using a GUI program for SFTP:
> 		Remember to explicitly specify into the program the port `2222`, where the `sftp` service is running.

Reviewer: ```suggestion
```

Perhaps remove this section if we want to assume port 22, which probably is the default?
